cheatsheets/Password_Storage_Cheat_Sheet.md

```diff
@@ -1,567 +1,180 @@
 
-2. Load and use new protection scheme
-    1. Load a new, stronger credential protection scheme (See next section on: Upgrading your existing password hashing solution)
-    2. Include version information stored with form
-    3. Set 'tainted'/'compromised' bit until user resets credentials
-    4. Rotate any keys and/or adjust protection function parameters such as work factor or salt
-    5. Increment scheme version number
+In some cases, it may be possible to increase the work factor of the hashes without the original password, although this is not supported by common hashing algorithms such as Bcrypt and PBKDF2.
 
-3. When user logs in:
-    1. Validate credentials based on stored version (old or new); if older compromised version is still active for user, demand 2nd factor or secret answers until the new method is implemented or activated for that user
-    2. Prompt user for credential change, apologize, & conduct out-of-band confirmation
-    3. Convert stored credentials to new scheme as user successfully log in
+## Maximum Password Lengths
 
-## Upgrading your existing password hashing solution
+Some hashing algorithms such as Bcrypt have a maximum length for the input, which is 72 characters for most implementations (there are some [reports](https://security.stackexchange.com/questions/39849/does-bcrypt-have-a-maximum-password-length) that other implementations have lower maximum lengths, but none have been identified at the time of writing). Where Bcrypt is used, a maximum length of 64 characters should be enforced on the input, as this provides a sufficiently high limit, while still allowing for string termination issues, and not revealing that the application uses Bcrypt.
 
-The above guidance describes how to do password hashing correctly/safely. However, it is very likely you'll be in a situation where you have an existing solution you want to upgrade. This [article](https://veggiespam.com/painless-password-hash-upgrades/) provides some good guidance on how to accomplish an upgrade in place without adversely affecting existing user accounts and future proofing your upgrade so you can seamlessly upgrade again (which you eventually will need to do).
+Additionally, due to how computationally expensive modern hashing functions are, if a user can supply very long passwords then there is a potential denial of service vulnerability, such as the one published in [Django](https://www.djangoproject.com/weblog/2013/sep/15/security/) in 2013.
 
-# Argon2 usage proposal in Java
+In order to protect against both of these issues, a maximum password length should be enforced. This should be 64 characters for Bcrypt (due to limitations in the algorithm and implementations), and between 64 and 128 characters for other algorithms. The easiest way to achieve this is a simple limit on the maximum length of the password.
```

Comment:
The last sentence doesn't fit properly in this paragraph. Sounds a bit redundant and have no meaning in saying it. I believe it was an old sentence?

Reply:
Removed